contracts/sol6/fprConversionRates/SimpleVolumeImbalanceRecorder.sol

```diff
@@ -0,0 +1,150 @@
+pragma solidity 0.6.6;
+
+import "@kyber.network/utils-sc/contracts/Utils.sol";
+import "@kyber.network/utils-sc/contracts/Withdrawable.sol";
+
+/// @dev this is another verison of SignedSafeMath(OpenZeppelin) for int64
+library SafeInt64 {
+    function add(int64 a, int64 b) internal pure returns (int64) {
+        int64 c = a + b;
+        require((b >= 0 && c >= a) || (b < 0 && c < a), "SafeInt64: addition overflow");
+        return c;
+    }
+}
+
+contract SimpleVolumeImbalanceRecorder is Withdrawable {
+    using SafeInt64 for int64;
+
+    int256 internal constant MAX_INT64 = 2**63 - 1;
+    int256 internal constant MIN_INT64 = -(2**63);
+    /// @dev data can be compressed into 1 word in storage
+    struct TokenControlInfo {
+        uint128 minimalRecordResolution; // should be roughly 1 cent
+        uint64 maxPerBlockImbalanceInResolution; // in resolution
+        uint64 maxTotalImbalanceInResolution; // in resolution
+    }
+
+    /// @dev data can be compressed into 1 word in storage
+    struct TokenImbalanceData {
+        int64 lastBlockBuyUnitsImbalance;
+        uint64 lastBlock;
+        int64 totalBuyUnitsImbalance;
+        uint64 lastRateUpdateBlock;
+    }
+
+    mapping(IERC20Ext => TokenControlInfo) internal tokenControlInfo;
+    mapping(IERC20Ext => TokenImbalanceData) public tokenImbalanceData;
+
+    constructor(address _admin) public Withdrawable(_admin) {}
+
+    function setTokenControlInfo(
+        IERC20Ext token,
+        uint128 minimalRecordResolution,
+        uint256 maxPerBlockImbalance,
+        uint256 maxTotalImbalance
+    ) external onlyAdmin {
+        require(minimalRecordResolution != 0, "zero minimalRecordResolution");
+        require(maxPerBlockImbalance != 0, "zero maxPerBlockImbalance");
+        require(maxTotalImbalance != 0, "zero maxTotalImbalance");
```

Comment:
I see the types changed from unit256
I am not sure we can have this change.
need to better analyse number limits and math for this.
also figure out what do we expect for minimalRecordResolution values.

Reply:
uin64 ~ 10^18
uint128 ~ 10^36
I believe this limit is enough for market maker

Reply:
Because imbalance data (in resolution) stores in this contract is only int64. 
So max value of it (in resolution) should not greater than max_int_64.

Anyway. I'll add documentation about this